File: Sources/Client/InputManager.cpp
#include "InputManager.h"
#include <cstring>

InputManager& InputManager::Get()
{
	static InputManager s_instance;
	return s_instance;
}

InputManager::InputManager()
	: m_hWnd(nullptr)
	, m_active(false)
	, m_mouseX(0)
	, m_mouseY(0)
	, m_wheelDelta(0)
	, m_leftDown(false)
	, m_rightDown(false)
	, m_leftPressed(false)
	, m_rightPressed(false)
	, m_leftReleased(false)
	, m_rightReleased(false)
{
	std::memset(m_keyDown, 0, sizeof(m_keyDown));
	std::memset(m_keyPressed, 0, sizeof(m_keyPressed));
	std::memset(m_keyReleased, 0, sizeof(m_keyReleased));
}

InputManager::~InputManager() = default;

void InputManager::Initialize(HWND hWnd)
{
	m_hWnd = hWnd;
	UpdateCursorClip(false);

	if (m_hWnd && GetForegroundWindow() == m_hWnd) {
		SetActive(true);
	}
}

void InputManager::SetActive(bool active)
{
	m_active = active;
	UpdateCursorClip(active);
	if (!active) {
		ClearAllKeys();
	}
}

void InputManager::BeginFrame()
{
	m_leftPressed = false;
	m_rightPressed = false;
	m_leftReleased = false;
	m_rightReleased = false;
	m_wheelDelta = 0;
	std::memset(m_keyPressed, 0, sizeof(m_keyPressed));
	std::memset(m_keyReleased, 0, sizeof(m_keyReleased));
}

void InputManager::OnMouseMove(int x, int y)
{
	UpdateLogicalPosition(x, y);
}

void InputManager::OnMouseDown(int button)
{
	if (button == 0) {
		if (!m_leftDown) {
			m_leftPressed = true;
		}
		m_leftDown = true;
	}
	else if (button == 1) {
		if (!m_rightDown) {
			m_rightPressed = true;
		}
		m_rightDown = true;
	}

	if (m_hWnd) {
		SetCapture(m_hWnd);
	}
}

void InputManager::OnMouseUp(int button)
{
	if (button == 0) {
		if (m_leftDown) {
			m_leftReleased = true;
		}
		m_leftDown = false;
	}
	else if (button == 1) {
		if (m_rightDown) {
			m_rightReleased = true;
		}
		m_rightDown = false;
	}

	if (!m_leftDown && !m_rightDown && GetCapture() == m_hWnd) {
		ReleaseCapture();
	}
}

void InputManager::OnMouseWheel(int delta, int x, int y)
{
	UpdateLogicalPosition(x, y);
	m_wheelDelta = static_cast<short>(m_wheelDelta + delta);
}

void InputManager::OnKeyDown(int vk)
{
	if (vk < 0 || vk >= kKeyCount) {
		return;
	}
	m_keyPressed[vk] = true;
	m_keyDown[vk] = true;
}

void InputManager::OnKeyUp(int vk)
{
	if (vk < 0 || vk >= kKeyCount) {
		return;
	}
	if (m_keyDown[vk]) {
		m_keyReleased[vk] = true;
	}
	m_keyDown[vk] = false;
}

bool InputManager::IsLeftMouseDown() const
{
	return m_leftDown;
}

bool InputManager::IsLeftMousePressed() const
{
	return m_leftPressed;
}

bool InputManager::IsLeftMouseReleased() const
{
	return m_leftReleased;
}

bool InputManager::IsRightMouseDown() const
{
	return m_rightDown;
}

bool InputManager::IsRightMousePressed() const
{
	return m_rightPressed;
}

bool InputManager::IsRightMouseReleased() const
{
	return m_rightReleased;
}

bool InputManager::IsKeyDown(int vk) const
{
	if (vk < 0 || vk >= kKeyCount) {
		return false;
	}
	return m_keyDown[vk];
}

bool InputManager::IsKeyPressed(int vk) const
{
	if (vk < 0 || vk >= kKeyCount) {
		return false;
	}
	return m_keyPressed[vk];
}

bool InputManager::IsKeyReleased(int vk) const
{
	if (vk < 0 || vk >= kKeyCount) {
		return false;
	}
	return m_keyReleased[vk];
}

bool InputManager::ConsumeKeyPressed(int vk)
{
	if (vk < 0 || vk >= kKeyCount) {
		return false;
	}
	if (!m_keyPressed[vk]) {
		return false;
	}
	m_keyPressed[vk] = false;
	return true;
}

void InputManager::ClearKeyPressed(int vk)
{
	if (vk < 0 || vk >= kKeyCount) {
		return;
	}
	m_keyPressed[vk] = false;
}

void InputManager::ClearAllKeys()
{
	std::memset(m_keyDown, 0, sizeof(m_keyDown));
	std::memset(m_keyPressed, 0, sizeof(m_keyPressed));
	std::memset(m_keyReleased, 0, sizeof(m_keyReleased));
}

bool InputManager::IsCtrlDown() const
{
	return IsKeyDown(VK_CONTROL);
}

bool InputManager::IsShiftDown() const
{
	return IsKeyDown(VK_SHIFT);
}

bool InputManager::IsAltDown() const
{
	return IsKeyDown(VK_MENU);
}

bool InputManager::IsEnterPressed() const
{
	return IsKeyPressed(VK_RETURN);
}

bool InputManager::IsEscPressed() const
{
	return IsKeyPressed(VK_ESCAPE);
}

void InputManager::SetEnterPressed()
{
	m_keyPressed[VK_RETURN] = true;
}

void InputManager::SetEscPressed()
{
	m_keyPressed[VK_ESCAPE] = true;
}

void InputManager::ClearEnterPressed()
{
	ClearKeyPressed(VK_RETURN);
}

void InputManager::ClearEscPressed()
{
	ClearKeyPressed(VK_ESCAPE);
}

short InputManager::GetMouseX() const
{
	return m_mouseX;
}

short InputManager::GetMouseY() const
{
	return m_mouseY;
}

short InputManager::GetWheelDelta() const
{
	return m_wheelDelta;
}

short InputManager::ConsumeWheelDelta()
{
	short delta = m_wheelDelta;
	m_wheelDelta = 0;
	return delta;
}

void InputManager::ClearWheelDelta()
{
	m_wheelDelta = 0;
}

void InputManager::SetMousePosition(short x, short y)
{
	m_mouseX = x;
	m_mouseY = y;
}

void InputManager::GetLegacyState(short* pX, short* pY, short* pZ, char* pLB, char* pRB) const
{
	if (pX) *pX = m_mouseX;
	if (pY) *pY = m_mouseY;
	if (pZ) *pZ = m_wheelDelta;
	if (pLB) *pLB = m_leftDown ? 1 : 0;
	if (pRB) *pRB = m_rightDown ? 1 : 0;
}

void InputManager::UpdateLogicalPosition(int clientX, int clientY)
{
	if (!m_hWnd) {
		m_mouseX = static_cast<short>(clientX);
		m_mouseY = static_cast<short>(clientY);
		return;
	}

	RECT rcClient{};
	GetClientRect(m_hWnd, &rcClient);
	int winW = rcClient.right - rcClient.left;
	int winH = rcClient.bottom - rcClient.top;
	if (winW <= 0 || winH <= 0) {
		return;
	}

	double scale = static_cast<double>(winW) / static_cast<double>(LOGICAL_WIDTH);
	double scaleY = static_cast<double>(winH) / static_cast<double>(LOGICAL_HEIGHT);
	if (scaleY < scale) {
		scale = scaleY;
	}
	if (scale <= 0.0) {
		scale = 1.0;
	}

	int destW = static_cast<int>(LOGICAL_WIDTH * scale);
	int destH = static_cast<int>(LOGICAL_HEIGHT * scale);
	int offsetX = (winW - destW) / 2;
	int offsetY = (winH - destH) / 2;

	long scaledX = static_cast<long>((clientX - offsetX) / scale);
	long scaledY = static_cast<long>((clientY - offsetY) / scale);

	if (scaledX < 0) scaledX = 0;
	if (scaledY < 0) scaledY = 0;
	if (scaledX > LOGICAL_MAX_X) scaledX = LOGICAL_MAX_X;
	if (scaledY > LOGICAL_MAX_Y) scaledY = LOGICAL_MAX_Y;

	m_mouseX = static_cast<short>(scaledX);
	m_mouseY = static_cast<short>(scaledY);
}

void InputManager::UpdateCursorClip(bool active)
{
	if (!m_hWnd) {
		return;
	}

	if (active) {
		RECT rcClient{};
		GetClientRect(m_hWnd, &rcClient);
		int winW = rcClient.right - rcClient.left;
		int winH = rcClient.bottom - rcClient.top;
		double scale = static_cast<double>(winW) / static_cast<double>(LOGICAL_WIDTH);
		double scaleY = static_cast<double>(winH) / static_cast<double>(LOGICAL_HEIGHT);
		if (scaleY < scale) {
			scale = scaleY;
		}
		if (scale <= 0.0) {
			scale = 1.0;
		}
		int destW = static_cast<int>(LOGICAL_WIDTH * scale);
		int destH = static_cast<int>(LOGICAL_HEIGHT * scale);
		int offsetX = (winW - destW) / 2;
		int offsetY = (winH - destH) / 2;

		POINT ptTopLeft{ offsetX, offsetY };
		POINT ptBottomRight{ offsetX + destW, offsetY + destH };
		ClientToScreen(m_hWnd, &ptTopLeft);
		ClientToScreen(m_hWnd, &ptBottomRight);
		RECT rcClip{ ptTopLeft.x, ptTopLeft.y, ptBottomRight.x, ptBottomRight.y };
		ClipCursor(&rcClip);
	}
	else {
		ClipCursor(nullptr);
	}
}

File: engine/src/core/InputManager.ts
/**
 * InputManager — Frame-based keyboard and mouse input singleton.
 * Mirrors C++ InputManager from Sources/Client/InputManager.h/.cpp.
 *
 * Tracks down/pressed/released per frame. Call beginFrame() once per frame
 * to reset pressed/released flags before processing new events.
 *
 * Uses string key names (KeyboardEvent.key.toLowerCase()) instead of
 * Win32 virtual-key codes. Mouse buttons: 0 = left, 2 = right.
 */

export class InputManager {
	private static s_instance: InputManager | null = null;

	static get(): InputManager {
		if (!InputManager.s_instance) {
			InputManager.s_instance = new InputManager();
		}
		return InputManager.s_instance;
	}

	// ── Keyboard state ────────────────────────────────────────────

	private m_keyDown = new Set<string>();
	private m_keyPressed = new Set<string>();
	private m_keyReleased = new Set<string>();

	// ── Mouse state ───────────────────────────────────────────────

	private m_mouseX = 0;
	private m_mouseY = 0;
	private m_wheelDelta = 0;

	private m_leftDown = false;
	private m_rightDown = false;
	private m_leftPressed = false;
	private m_rightPressed = false;
	private m_leftReleased = false;
	private m_rightReleased = false;

	// ── Modifier cache ────────────────────────────────────────────

	private m_shiftDown = false;
	private m_ctrlDown = false;
	private m_altDown = false;

	// ── Active flag ───────────────────────────────────────────────

	private m_active = true;

	// ── Initialization ────────────────────────────────────────────

	/**
	 * Attach DOM event listeners to the given element (usually the canvas
	 * or document). Call once during CGame.bInit().
	 */
	initialize(target: HTMLElement | Document = document): void {
		target.addEventListener('keydown', this.onKeyDown);
		target.addEventListener('keyup', this.onKeyUp);
		target.addEventListener('mousemove', this.onMouseMove);
		target.addEventListener('mousedown', this.onMouseDown);
		target.addEventListener('mouseup', this.onMouseUp);
		target.addEventListener('wheel', this.onWheel, { passive: false });
		target.addEventListener('contextmenu', (e) => e.preventDefault());

		// Track window blur to reset all keys (prevents sticky keys)
		window.addEventListener('blur', this.onBlur);
	}

	setActive(active: boolean): void {
		this.m_active = active;
	}

	/** Reset per-frame pressed/released flags. Call once per frame. */
	beginFrame(): void {
		this.m_keyPressed.clear();
		this.m_keyReleased.clear();
		this.m_leftPressed = false;
		this.m_rightPressed = false;
		this.m_leftReleased = false;
		this.m_rightReleased = false;
		this.m_wheelDelta = 0;
	}

	// ── Keyboard queries ──────────────────────────────────────────

	/** True while the key is held down. */
	isKeyDown(key: string): boolean {
		return this.m_keyDown.has(key.toLowerCase());
	}

	/** True only on the frame the key was first pressed. */
	isKeyPressed(key: string): boolean {
		return this.m_keyPressed.has(key.toLowerCase());
	}

	/** True only on the frame the key was released. */
	isKeyReleased(key: string): boolean {
		return this.m_keyReleased.has(key.toLowerCase());
	}

	/** Like isKeyPressed but consumes the event so subsequent checks return false. */
	consumeKeyPressed(key: string): boolean {
		const k = key.toLowerCase();
		if (this.m_keyPressed.has(k)) {
			this.m_keyPressed.delete(k);
			return true;
		}
		return false;
	}

	clearKeyPressed(key: string): void {
		this.m_keyPressed.delete(key.toLowerCase());
	}

	clearAllKeys(): void {
		this.m_keyDown.clear();
		this.m_keyPressed.clear();
		this.m_keyReleased.clear();
		this.m_shiftDown = false;
		this.m_ctrlDown = false;
		this.m_altDown = false;
	}

	// ── Modifier queries ──────────────────────────────────────────

	isShiftDown(): boolean { return this.m_shiftDown; }
	isCtrlDown(): boolean { return this.m_ctrlDown; }
	isAltDown(): boolean { return this.m_altDown; }

	// ── Special key shortcuts (mirrors C++ convenience methods) ───

	isEnterPressed(): boolean { return this.isKeyPressed('enter'); }
	isEscPressed(): boolean { return this.isKeyPressed('escape'); }

	// ── Mouse queries ─────────────────────────────────────────────

	getMouseX(): number { return this.m_mouseX; }
	getMouseY(): number { return this.m_mouseY; }
	getWheelDelta(): number { return this.m_wheelDelta; }

	consumeWheelDelta(): number {
		const d = this.m_wheelDelta;
		this.m_wheelDelta = 0;
		return d;
	}

	setMousePosition(x: number, y: number): void {
		this.m_mouseX = x;
		this.m_mouseY = y;
	}

	isLeftMouseDown(): boolean { return this.m_leftDown; }
	isLeftMousePressed(): boolean { return this.m_leftPressed; }
	isLeftMouseReleased(): boolean { return this.m_leftReleased; }

	isRightMouseDown(): boolean { return this.m_rightDown; }
	isRightMousePressed(): boolean { return this.m_rightPressed; }
	isRightMouseReleased(): boolean { return this.m_rightReleased; }

	// ── Legacy state (mirrors C++ GetLegacyState) ─────────────────

	getLegacyState(): { x: number; y: number; z: number; lb: number; rb: number } {
		return {
			x: this.m_mouseX,
			y: this.m_mouseY,
			z: this.m_wheelDelta,
			lb: this.m_leftDown ? 1 : 0,
			rb: this.m_rightDown ? 1 : 0,
		};
	}

	// ── DOM event handlers (arrow functions for correct `this`) ───

	private onKeyDown = (e: Event): void => {
		if (!this.m_active) return;
		const ke = e as KeyboardEvent;
		const key = ke.key.toLowerCase();

		this.m_shiftDown = ke.shiftKey;
		this.m_ctrlDown = ke.ctrlKey;
		this.m_altDown = ke.altKey;

		if (!this.m_keyDown.has(key)) {
			this.m_keyDown.add(key);
			this.m_keyPressed.add(key);
		}
	};

	private onKeyUp = (e: Event): void => {
		if (!this.m_active) return;
		const ke = e as KeyboardEvent;
		const key = ke.key.toLowerCase();

		this.m_shiftDown = ke.shiftKey;
		this.m_ctrlDown = ke.ctrlKey;
		this.m_altDown = ke.altKey;

		this.m_keyDown.delete(key);
		this.m_keyReleased.add(key);
	};

	private onMouseMove = (e: Event): void => {
		if (!this.m_active) return;
		const me = e as MouseEvent;
		this.m_mouseX = me.offsetX;
		this.m_mouseY = me.offsetY;
	};

	private onMouseDown = (e: Event): void => {
		if (!this.m_active) return;
		const me = e as MouseEvent;
		if (me.button === 0) { this.m_leftDown = true; this.m_leftPressed = true; }
		if (me.button === 2) { this.m_rightDown = true; this.m_rightPressed = true; }
	};

	private onMouseUp = (e: Event): void => {
		if (!this.m_active) return;
		const me = e as MouseEvent;
		if (me.button === 0) { this.m_leftDown = false; this.m_leftReleased = true; }
		if (me.button === 2) { this.m_rightDown = false; this.m_rightReleased = true; }
	};

	private onWheel = (e: Event): void => {
		if (!this.m_active) return;
		const we = e as WheelEvent;
		we.preventDefault();
		this.m_wheelDelta += we.deltaY > 0 ? -1 : we.deltaY < 0 ? 1 : 0;
	};

	private onBlur = (): void => {
		this.clearAllKeys();
		this.m_leftDown = false;
		this.m_rightDown = false;
	};
}
